### DISCUSSION TITLE
Editoast, chartos and api merge proposal

### DISCUSSION THREAD

**Discussion by multun** [OpenRailAssociation/osrd#1782] [Ideas]
# Description of the new architecture

![architecture-merged](https://user-images.githubusercontent.com/5047140/187683954-7238a2a4-ae3f-49da-b481-19bf7507252b.png)

The only change to the previous architecture is the merge of chartos, editoast and the api into a single rust service.

# Upsides

- there would be single web service queried by the front-end, unifying chartos, editoast and the django api service.
- the postgres database wont be shared by multiple services anymore
- no more invalidating the chartos cache from editoast
- easier refactorings across codebases
- less tech stacks: we would switch from django + fastapi + java + rust to just java + rust
- speedup all kind of processing currently done by the API, such as post-processing simulation results
- consistent policies / authentication are easier to implement

# Drawbacks

- migrating costs time
- no more django admin
- rust ORMs aren't yet as fancy as the django ORM
- no more django rest framework
- rust has a steeper learning curve than python

# Migration plan

- until the plan starts, avoid adding new code into the django api
- rename editoast to **FIXME**
- port chartos to rust, merge it into **FIXME**
- port api to rust, merge it into **FIXME**
  - migrate all non-trivial views, such as pathfinding and simulation to **FIXME**
  - move the migration and database ownership to **FIXME**

**Comment by flomonster** [comment]
This has been done
